core/state/statedb.go

```diff
@@ -486,7 +486,14 @@ func (db *StateDB) ForEachStorage(addr common.Address, cb func(key, value common
 			cb(key, value)
 			continue
 		}
-		cb(key, common.BytesToHash(it.Value))
+
+		if len(it.Value) > 0 {
+			_, content, _, err := rlp.Split(it.Value)
```

Comment:
Why do you need to post-process the value returned by the iterator? Shouldn't the underlying iterator do this?

Reply:
Yes it is quite weird, since before putting the value into trie, we have an additional encode process https://github.com/ethereum/go-ethereum/blob/master/core/state/state_object.go#L234
And for the normal GetState operation, there is a corresponding decode process https://github.com/ethereum/go-ethereum/blob/master/core/state/state_object.go#L186

While the iterator only iterate the trie node, so we should also decode in the state iterator